src/jit/importer.cpp

```diff
@@ -4105,6 +4105,13 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_GC_KeepAlive:
+            {
+                retNode = gtNewOperNode(GT_SIDEEFFECT, TYP_VOID, impPopStack().val);
+                retNode->gtFlags |= GTF_CALL; // TODO: Is this the right side-effect flag to use?
```

Comment:
This is going to be a bit more conservative than needed but I don't think it matters, this node is unlikely to appear in call arguments, where it would matter. `GTF_GLOB_REF` should be added to, it normally comes with real calls. I don't think it makes sense in the case of `KeepAlive` but we need to be sure that no reordering occurs.

Reply:
Thanks! Fixed

Reply:
Hm, another approach could be not to set any flags here (because that node doesn't call anything) and make sure that the node can't be deleted, it means add it here:
https://github.com/dotnet/coreclr/blob/41546b05e31c0ba72537ec6445176086964717bb/src/jit/liveness.cpp#L2014-L2020

so the node won't be deleted.

It won't be moved across other calls because its operand should be a lclVar with `lvAddrExposed` flag set and for  such local vars we don't do movements.

The only exception that I can think of right now is unnecessary using of  `GC::KeepAlive( a);`:
```
Object a = new Object();
foo();
GC::KeepAlive(a);
```
LclVar for `a` won't be marked as address exposed (because it is not) and instruction for `GC::KeepAlive(a);` in this case can be moved before `foo();`, but it doesn't look like a problem, does it?

Also, I am not sure if we do any instruction reordering like this between different statements/blocks at all.

Reply:
>  its operand should be a lclVar with lvAddrExposed flag

The operand for `GC.KeepAlive` can be anything. It does not need to be  `lclVar` with `lvAddrExposed` flag. It is actually pretty rare for the operand to be `lvAddrExposed`.

> The only exception that I can think of right now is unnecessary using of GC::KeepAlive

Your example is a perfectly valid use and useful use of `GC.KeepAlive`.

Reply:
> in this case can be moved before foo();, but it doesn't look like a problem, does it?

Moving the GC.KeepAlive in before foo would be invalid optimization.

Reply:
Is it necessary for the operand of `GC.KeepAlive` to be in memory? If not, it would be overkill to mark it with `lvAddrExposed`. 
Although we don't currently reorder instructions, I think we should ensure that we've modeled this in a way that won't create problems in future. It seems to me that the right semantic for this is that it has global effect (`GTF_GLOB_REF`) and not mark it as if it's a call. We should never, even in  future, reorder things with this flag set.

Reply:
I do not believe `GTF_GLOB_REF` alone is enough to prevent reordering in general. 

It should work in the case of `GT_SIDEEFFECT`, it's `TYP_VOID` so the importer has to make a statement out of it. `gtSetEvalOrder` can't touch it since it's the root of the tree and CSE/loop hoisting can't touch it either because of `TYP_VOID`. Well, at least currently.

Otherwise, in the absence of `GTF_ASG`/`GTF_CALL`, `GTF_GLOB_REF` means that the tree contains aliased loads and loads can and do reorder (or get eliminated).

Though apparently `gtCanSwapOrder` gets it wrong and checks for `GTF_GLOB_EFFECT` instead of `GTF_SIDE_EFFECT`. Oh well, I do not trust side effect flags too much anyway. It does check for `GTF_ORDER_SIDEEFF` but then I don't remember loop hoisiting checking for that so... maybe it does work, maybe it does not.

Reply:
I think you have convinced me that it is better to be more conservative with the side-effects. I have flipped it back to include `GTF_CALL` and added comment.